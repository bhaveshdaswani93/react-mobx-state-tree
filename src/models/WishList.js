import { types,getParent,destroy } from 'mobx-state-tree';

export const WishListItem = types.model({
    name:types.string,
    price:types.number,
    image:types.optional(types.string,"")
})

.actions(self=>({
    changeName(newName) {
        self.name = newName;
    },
    changePrice(newPrice) {
        self.price = newPrice
    },
    changeImage(newImage){
        self.image = newImage
    },
    delete(){
        getParent(self,2).remove(self)
    },
    beforeDestroy(){
        alert('Wishlist item deleted'+JSON.stringify(self));
    }
}))

export const WishList = types.model({
    items: types.optional( types.array(WishListItem),[])
})
.actions(self=>({
    add(item) {
        self.items.push(item);
    },
    remove(item) {
        destroy(item)
        // self.items.splice(self.items.indexOf(item),1)
    }
}))
.views(self=>({
    get totalPrice(){
        return self.items.reduce((sum,item)=>sum+item.price,0)
    }
}))
;