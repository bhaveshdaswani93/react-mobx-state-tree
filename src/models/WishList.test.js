import { getSnapshot,onSnapshot,onPatch } from 'mobx-state-tree';
import { reaction } from 'mobx';

import { WishListItem,WishList } from './WishList';

it('can create Wishlist item',()=>{
    const item = WishListItem.create({
        "name":"demo",
        "price":3
    })
    expect(item.name).toBe('demo')
    // item.name = 'test';
    item.changeName('test');
    item.changePrice(9)
    expect(item.name).toBe('test');
    expect(item.price).toBe(9)


    

    
})

it('can create wishlist',()=>{
    const WistListItems = WishList.create({
        items:[
            {
                "name":"demo",
                "price":3
            }
        ]
    })
    
    expect(WistListItems.items.length).toBe(1);
    expect(WistListItems.items[0].price).toBe(3);
    expect(WistListItems.items[0].image).toBe('');

})

it('can add item to wistlist',()=>{
    const wistlist = WishList.create();
    const states = [];
    onSnapshot(wistlist,snapshot=>{
        states.push(snapshot);
    })
    wistlist.add({
        "name":"demo",
        "price":3
    });
    expect(wistlist.items.length).toBe(1);
    expect(wistlist.items[0].price).toBe(3);
    wistlist.items[0].changeName('change this name');
    expect(wistlist.items[0].name).toBe('change this name');
    expect(getSnapshot(wistlist)).toMatchSnapshot()

    wistlist.items[0].changePrice(21);
    expect(wistlist.items[0].price).toBe(21);

    expect(states).toMatchSnapshot();

})


it('can record patch',()=>{
    const list = WishList.create();
    const patches = [];
    onPatch(list,patch=>{
        patches.push(patch)
    })

    list.add({
        name:'hello',
        price:90
    })
    list.items[0].changeName('good');
    list.items[0].changePrice(87);
    expect(patches).toMatchSnapshot();
})

it('can calculate total price',()=>{
    const list = WishList.create({
        items:[
            {
                name:'First',
                price:19
            },
            {
                name:'second',
                price:61
            }
        ]
    })
    expect(list.totalPrice).toBe(80);
    let count=0;
    reaction(()=>list.totalPrice,()=>count++);
    expect(count).toBe(0);
    list.items[0].changeName('New Name');
    expect(count).toBe(0);
    list.items[0].changePrice(21);
    expect(count).toBe(1);
    expect(list.totalPrice).toBe(82);
})