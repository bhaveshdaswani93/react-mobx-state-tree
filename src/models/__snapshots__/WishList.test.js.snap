// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add item to wistlist 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "change this name",
      "price": 3,
    },
  ],
}
`;

exports[`can add item to wistlist 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "demo",
        "price": 3,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "change this name",
        "price": 3,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "change this name",
        "price": 21,
      },
    ],
  },
]
`;

exports[`can record patch 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "hello",
      "price": 90,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "good",
  },
  Object {
    "op": "replace",
    "path": "/items/0/price",
    "value": 87,
  },
]
`;
